### PR TITLE
Themes SSR: Rely on Redux store for caching themeDetails API object

### PR DESCRIPTION
**Pull request by ockham**
Simplify the caching logic by relying on the `themeDetails` Redux subtree to cache theme details API data.

Since the theme object is part of `initialState` which is used to compute the
markup caching key, an updated theme should still cause a markup cache miss and
hence a freshly rendered page. This way, the functionality provided by the API
data cache is retained.

Also, markup is now returned on first hit (when cache is empty), mostly for
better DevX.

To test -- verify that everyting still works as before, except for theme details pages being SSR'ed on first hit already.

### CODE REVIEW

File: client/my-sites/themes/controller/index.node.js [hunk 1]
```diff
@@ -43,27 +38,8 @@ export function fetchThemeDetailsData( context, next ) {
...
-		}
-		const themeData = themeDetailsCache.get( themeSlug );
-		if ( ! themeData || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
-			debug( 'caching', themeSlug );
-			themeDetailsCache.set( themeSlug, data );
-			context.store.dispatch( receiveThemeDetails( data ) );
-			next();
-		}
-	} );
+	themeSlug && context.store.dispatch( fetchThemeDetails( themeSlug, next ) );
```
Comment (seear) [inline client/my-sites/themes/controller/index.node.js]: Does this mean that we don't move on to the render step until the API response is received?

Comment (ockham) [inline client/my-sites/themes/controller/index.node.js]: Yeah, that's right. So if we want to avoid that, we could re-enable some sort of caching. As I said, my preferred approach would be encapsulating it in our Redux tree. We could introduce a `getThemeDetails()` selector and use that in lieu of L46-L52. The reducer is already storing theme details per-theme-id IIRC, so it might work OOTBish...

**Comment by seear** [comment]
> So if we want to avoid that, we could re-enable some sort of caching. As I said, my preferred approach would be encapsulating it in our Redux tree.

Feel free to give that a try, because I think that would be a really tidy solution. Let's not merge this until we have either replaced the cache or taken some measurements that show that API response times at the node servers are fast enough not to worry about.

**Comment by ockham** [comment]
Pushed a commit to use the selector, which was fortunately really easy. Not quite sure how to best test if it's used.

(This should also pick up re-hydrated `themeDetails` from `initialReduxState` on the client-side, which is a nice side effect.)

File: client/my-sites/themes/controller/index.node.js [hunk 2]
```diff
@@ -42,27 +40,14 @@ export function fetchThemeDetailsData( context, next ) {
 	}
 
 	const themeSlug = context.params.slug;
-	const theme = themeDetailsCache.get( themeSlug );
+	const theme = getThemeDetails( context.store.getState(), themeSlug );
 
-	if ( theme ) {
+	if ( Object.keys( theme ).length ) {
```
Comment (seear) [inline client/my-sites/themes/controller/index.node.js]: Should we give the theme a time-to-live, or maybe do something with `date_updated`? As it stands this will rely on a Calypso re-deploy to flush the cache.

Comment (seear) [inline client/my-sites/themes/controller/index.node.js]: Alternatively, we could add a call to `fetchThemeDetails` in this branch (without the callback), so that the cache is always up to date.

Comment (ockham) [inline client/my-sites/themes/controller/index.node.js]: Since we now rely on the Redux store for caching (which is initialized on a per-request basis [here](https://github.com/Automattic/wp-calypso/blob/master/server/isomorphic-routing/index.js#L44)), I'm thinking that the cache probably doesn't even live longer than a request. Which kinda devaluates the whole thing. :-/ (BTW I think it wasn't quite working before either, since `if ( theme )` yields `true` for an empty object.)

Comment (ockham) [inline client/my-sites/themes/controller/index.node.js]: Noting here that we discussed in Hangouts that we could try to persist cache across requests by creating the Redux store at module level (in `server/isomorphic-rendering`, much [like we did with `Map`](https://github.com/Automattic/wp-calypso/pull/3774/files#diff-3bf7293097114d293852f6d4c4d97aa7L24) up until this PR, since we don't initialize any other Redux subtrees on the server side yet.

However, I'm starting to think that this doesn't make much of a diference, and that instead, the Redux store would still be re-created on each new import of that module -- and that it's the same with the `Map` based cache now. My evidence is that after browsing two or more different theme details pages on `wpcalypso`, the `initialReduxStore` var in the page source always only holds `themeDetails` for the current theme.

Comment (ockham) [inline client/my-sites/themes/controller/index.node.js]: To move forward, I'd suggest merging this PR so we get at least the benefit

> This should also pick up re-hydrated `themeDetail` from `initialReduxState` on the client-side, which is a nice side effect.

and some more easily understandable code. Persisting theme details API cache across requests can be the subject of a separate PR.

We should however keep in mind that having `long_description`s for ~300 themes in `initialReduxState` would blow up our initial HTML really badly!

**Comment by seear** [comment]
> and that it's the same with the Map based cache now.

@ockham: are you saying that the current map-based cache does not work?

**Comment by ockham** [comment]
I guess I am -- how can we know?

**Comment by seear** [comment]
Logging the map size (on master)—seems fine:
```
Ready! You can load http://calypso.localhost:3000/ now. Have fun!
theme cache size: 0
GET /theme/mood 200 776.937 ms - 4350
GET /calypso/style-debug.css?v=981c2c6159 304 8.330 ms - -
HEAD /version?1459514809806 200 3.162 ms - 20
GET /calypso/images/favicons/favicon-development.ico 304 1.826 ms - -
theme cache size: 1
GET /theme/mood 200 441.588 ms - 48832
GET /calypso/style-debug.css?v=981c2c6159 304 2.401 ms - -
HEAD /version?1459514824237 200 0.935 ms - 20
theme cache size: 1
GET /theme/twentysixteen 200 297.046 ms - 4350
GET /calypso/style-debug.css?v=981c2c6159 304 3.115 ms - -
HEAD /version?1459514830191 200 0.847 ms - 20
GET /calypso/images/favicons/favicon-development.ico 304 1.107 ms - -
theme cache size: 2
GET /theme/twentyfifteen 200 248.861 ms - 4350
GET /calypso/style-debug.css?v=981c2c6159 304 3.532 ms - -
HEAD /version?1459514837954 200 0.748 ms - 20
theme cache size: 3
GET /theme/twentyfifteen 200 309.057 ms - 53443
GET /calypso/style-debug.css?v=981c2c6159 304 3.338 ms - -
```

**Comment by ockham** [comment]
Noting there that we've decided to try using the REST API sync handler (which caches API responses for use in offline Calypso) here. See e.g. #2843.

**Comment by ockham** [comment]
> Noting there that we've decided to try using the REST API sync handler (which caches API responses for use in offline Calypso) here. See e.g. #2843.

Hmm, that relies on `localForage`. That might be client-side only then :-(

**Comment by lancewillett** [comment]
Closing this as it's been inactive for several months. Feel free to resurrect it in the future if it's needed or important.

**Comment by ockham** [comment]
Essentially fixed by https://github.com/Automattic/wp-calypso/pull/13339.
